### PR TITLE
ch4: refactor spawn functions

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Revamp the current `ofi_spawn.c` with netmod abstraction so the function can be easily implemented in other netmod and avoid code duplications.

With this revamp, the only netmod dependent routines are `dynamic_send` and `dynamic_recv`. With proper API exposure, the rest of the routines in `ofi_spawn.c` can be moved to upper layer. For ucx to support spawn, all it needs are the two dynamic send/recv functions.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_spawn.c [hunk 1]
```diff
@@ -509,7 +509,7 @@ int MPIDI_OFI_upids_to_lupids(int size, size_t * remote_upid_size, char *remote_
             MPIDI_OFI_AV(&MPIDIU_get_av(avtid, i)).dest[nic][0] = addr;
             /* highest bit is marked as 1 to indicate this is a new process */
             (*remote_lupids)[new_avt_procs[i]] = MPIDIU_LUPID_CREATE(avtid, i);
-            MPIDIU_LUPID_SET_NEW_AVT_MARK((*remote_lupids)[new_avt_procs[i]]);
+            // MPIDIU_LUPID_SET_NEW_AVT_MARK((*remote_lupids)[new_avt_procs[i]]);
```
Comment (yfguo) [inline src/mpid/ch4/netmod/ofi/ofi_spawn.c:512]: Is this removed or commented out?

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_spawn.c:512]: It is already removed in main. It will disappear once I rebase.

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Review by yfguo — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Pull request merged**